slides/02_r_packages.qmd

```diff
@@ -7,6 +7,310 @@ description: description. Photo CC0 by cottonbro studio on pexels.com.
+
+- We can include the documentation source directly in the R script, on top of the objects we are documenting
+- Syntax is composed of special comments `#'` and special macros preceded with `@`
+- In RStudio running Build > More > Document will render the `.Rd` files and the `NAMESPACE` file for you
+- Get started with `usethis::use_roxygen_md()`
+- Inside a function, click Code > Insert Roxygen Skeleton 
+
+## Setting up `roxygen2` in your project
+
+![](resources/roxygen2-setting.png){.r-stretch}
+
+## `roxygen2` Source
+
+`R/my_sum.R`:
+
+```{r}
+#| echo: true
+#| eval: false
+#' My Summation Function
+#'
+#' This is my first function and it sums two numbers.
+#'
+#' @param x first summand.
+#' @param y second summand.
+#'
+#' @return The sum of `x` and `y`.
+#' @export
+#' 
+#' @note This function is a bit boring but that is ok.
+#' @seealso [Arithmetic] for an easier way.
+#'
+#' @examples
+#' my_sum(1, 2)
+my_sum <- function(x, y) {
+  x + y
+}
+```
+
+## `roxygen2` Output
+
+`man/my_sum.Rd`:
+
+```latex
+% Generated by roxygen2: do not edit by hand
+% Please edit documentation in R/bla.R
+\name{my_sum}
+\alias{my_sum}
+\title{My Summation Function}
+\usage{
+my_sum(x, y)
+}
+\arguments{
+\item{x}{first summand.}
+
+\item{y}{second summand.}
+}
+\value{
+The sum of \code{x} and \code{y}.
+}
+\description{
+This is my first function and it sums two numbers.
+}
+\note{
+This function is a bit boring but that is ok.
+}
+\examples{
+my_sum(1, 2)
+}
+\seealso{
+\link{Arithmetic} for an easier way.
+}
+```
+
+## `roxygen2` Output (cont'd)
+
+`NAMESPACE`:
+
+```r
+# Generated by roxygen2: do not edit by hand
+
+export(my_sum)
+```
+
+## `tests` Folder
+
+- Here we store the unit tests covering the functionality of the package
+- Get started with `usethis::use_testthat()` and `usethis::use_test()`
+  and populate `tests/testthat` folder with unit tests
+- Rarely, tests cannot be run within `testthat` framework, then these can go
+  into R scripts directly in `tests` directory
+- We will look at unit tests in detail later
+
+## `data` Folder
+
+- For (example) data that you ship in your package to the user
+  - Get started with `usethis::use_data()`
+  - Note: Usually we use lazy data loading, therefore no `data()` call needed
+    before using the data
+- If you generate the example data, save the R script, too
+  - Put that into `data-raw` folder, start with `usethis::use_data_raw()`
+
+## `inst` Folder
+
+- Contents will be copied recursively to installation directory
+  - Be careful not to interfere with standard folder names
+- For data that is used by functions in the package itself
+  - Those would typically go into `inst/extdata` folder
+  - Load with `system.file("path/file", package = "mypackage")`  
+- `CITATION`: For custom `citation()` output
+  - Create it with `usethis::use_citation()`
+- `inst/doc` can contain documentation files (typically `pdf`)
+
+## `src` Folder
+
+- Contains sources and headers for any compiled code
```

Comment:
Maybe better because the code is not compiled yet:
- Contains sources and headers for any code that needs compilation